util/src/journaldb.rs

```diff
@@ -109,250 +154,136 @@ impl JournalDB {
 			r.append(id);
-
-			// Process the new inserts.
-			// We use the inserts for three things. For each:
-			// - we place into the backing DB or increment the counter if already in;
-			// - we note in the backing db that it was already in;
-			// - we write the key into our journal for this block;
-
-			r.begin_list(inserts.len());
-			inserts.iter().foreach(|&(k, _)| {r.append(&k);});
-			r.append(&removes);
-			Self::insert_keys(&inserts, &self.backing, &mut counters, &batch);
-			try!(batch.put(&last, r.as_raw()));
+			r.begin_list(inserted_keys.len());
+			for (k, v) in insertions {
+				r.begin_list(2);
+				r.append(&k);
+				r.append(&v);
+				journal_overlay.backing_overlay.emplace(k, v);
+			}
+			r.append(&removed_keys);
+
+			let mut k = RlpStream::new_list(3);
+			let index = journal_overlay.journal.get(&now).map(|j| j.len()).unwrap_or(0);
+			k.append(&now);
+			k.append(&index);
+			k.append(&&PADDING[..]);
+			try!(batch.put(&k.drain(), r.as_raw()));
 			try!(batch.put(&LATEST_ERA_KEY, &encode(&now)));
+			journal_overlay.journal.entry(now).or_insert_with(Vec::new).push(JournalEntry { id: id.clone(), insertions: inserted_keys, deletions: removed_keys });
 		}
 
+		let journal_overlay = journal_overlay.deref_mut();
 		// apply old commits' details
 		if let Some((end_era, canon_id)) = end {
-			let mut index = 0usize;
-			let mut last;
-			while let Some(rlp_data) = try!(self.backing.get({
-				let mut r = RlpStream::new_list(3);
-				r.append(&end_era);
-				r.append(&index);
-				r.append(&&PADDING[..]);
-				last = r.drain();
-				&last
-			})) {
-				let rlp = Rlp::new(&rlp_data);
-				let inserts: Vec<H256> = rlp.val_at(1);
-				let deletes: Vec<H256> = rlp.val_at(2);
-				// Collect keys to be removed. These are removed keys for canonical block, inserted for non-canonical
-				Self::kill_keys(if canon_id == rlp.val_at(0) {deletes} else {inserts}, &mut counters, &batch);
-				try!(batch.delete(&last));
-				index += 1;
+			if let Some(ref mut records) = journal_overlay.journal.get_mut(&end_era) {
+				let mut canon_insertions: Vec<(H256, Bytes)> = Vec::new();
+				let mut canon_deletions: Vec<H256> = Vec::new();
+				let mut overlay_deletions: Vec<H256> = Vec::new();
+				let mut index = 0usize;
+				for mut journal in records.drain(..) {
+					//delete the record from the db
+					let mut r = RlpStream::new_list(3);
+					r.append(&end_era);
+					r.append(&index);
+					r.append(&&PADDING[..]);
+					try!(batch.delete(&r.drain()));
+					trace!("commit: Delete journal for time #{}.{}: {}, (canon was {}): +{} -{} entries", end_era, index, journal.id, canon_id, journal.insertions.len(), journal.deletions.len());
+					{
+						if canon_id == journal.id {
+							for h in &journal.insertions {
+								match journal_overlay.backing_overlay.raw(&h) {
+									Some(&(ref d, rc)) if rc > 0 => canon_insertions.push((h.clone(), d.clone())), //TODO: optimizie this to avoid data copy
+									_ => ()
```

Comment:
`_ => ()` can be omitted

```rust
if Some(&(ref d, rc)) =  journal_overlay.backing_overlay.raw(h) {
  if rc > 0 {
    canon_insertions.push((h.clone(), d.clone())), //TODO: optimizie this to avoid data copy
  }
}
```